src/workloads/query/BooleanSimplifier.yml

```diff
@@ -0,0 +1,222 @@
+      Active: [4]
+      NopInPhasesUpTo: *MaxPhases
+      PhaseConfig:
+        Duration: *Duration
+        Collection: *Coll
+        Operations:
+        - OperationName: find
+          OperationCommand:
+            Filter:  {a2: 1, $or: [{c2: 1, d2: 2, b2: 3}, {c2: 1, d2: 2, e2: 4}]}
+
+# The simplifier simplifes the filter expression to {a3: 0, b3: 0} which makes evaluating the Filter
+# stage of Collection Scan plan mach cheaper.
+- Name: SimplifiedCollScanFilter.BM
+  Type: CrudActor
+  Database: *Database
+  Threads: *Threads
+  Phases:
+    OnlyActiveInPhases:
+      Active: [5]
+      NopInPhasesUpTo: *MaxPhases
+      PhaseConfig:
+        Duration: *Duration
+        Collection: *Coll
+        Operations:
+        - OperationName: find
+          OperationCommand:
+            Filter:  {
+                $and: [
+                    {b3: 0},
+                    {
+                        $or: [
+                            {a3: 0},
+                            {
+                                $and: [
+                                    {c3: {$in: *QueryStringIn}},
+                                    {c3: {$nin: *QueryStringIn}},
+                                ]
+                            }
+                        ]
+                    },
+                ]
+            }
+
+# The simplifier simplifes the filter expression to {b2: 3, c2: 3, d2: 3} which makes evaluating
+# the Filter stage of Index Scan plan mach cheaper.
+- Name: SimplifiedIndexScanFilter.BM
+  Type: CrudActor
+  Database: *Database
+  Threads: *Threads
+  Phases:
+    OnlyActiveInPhases:
+      Active: [6]
+      NopInPhasesUpTo: *MaxPhases
+      PhaseConfig:
+        Duration: *Duration
+        Collection: *Coll
+        Operations:
+        - OperationName: find
+          OperationCommand:
+            Filter:  {
+                $and: [
+                    {b2: 3},
+                    {c2: 3},
+                    {
+                        $or: [
+                            {d2: 3},
+                            {
+                                $and: [
+                                    {c3: {$in: *QueryStringIn}},
+                                    {c3: {$nin: *QueryStringIn}},
+                                ]
+                            }
+                        ]
+                    },
+                ]
+            }
+
+# The simplifier reduces the filter expression to {c2: 1, d2: 1} which makes passible a covered
+# index scan plan.
+- Name: CoveredQuery.BM
+  Type: CrudActor
+  Database: *Database
+  Threads: *Threads
+  Phases:
+    OnlyActiveInPhases:
+      Active: [7]
+      NopInPhasesUpTo: *MaxPhases
+      PhaseConfig:
+        Duration: *Duration
+        Collection: *Coll
+        Operations:
+        - OperationName: find
+          OperationCommand:
+            Filter:  {$or: [{c3: "HrHCEtBAGnv", c2: 1, d2: 1}, {c3: {$ne: "HrHCEtBAGnv"}, c2: 1, d2: 1}]}
+            Projection: {_id: 0, c2: 1, d2: 1}
```

Comment:
Could we add a test which uses a larger but also simplifiable predicate (I'm thinking in the range of 100+ clauses- here's an example of a [template](https://github.com/mongodb/genny/blob/28d4e7b8e7ca2171d260b17e92b45d78542b2f8d/src/workloads/query/FilterWithComplexLogicalExpression.yml#L58-L72 which generates an arbitrary number of clauses that would be helpful here)?

If we could add this test before index creation, we could also add the bonsai variants to AutoRun/ and could tag the data generated by that workload for the dashboard. This would allow us to see how Bonsai differs against master after your change/ if we need to address any difference here sooner rather than later.

I'd also say that for that particular workload, we could have a small collection & a large collection for the same workload- so, first we would create the small collection, say 100 or 500 docs, run this new workload, and then big collection with 100k docs, run the same workload again, then create indexes, and then run the remaining workloads on the big collection.

CC @timourk since this is relevant to the discussion yesterday around having two sizes of collection in workloads.

Reply:
I am afraid I cannot add these tests right now, because the simplifier is enabled only for some restricted cases:

- because of performance we wouldn't even attempt to simplify a complex expression at the moment, this can be changed in the future
- it is disabled for Bonsai. Because, Bonsai tests are flaky in general. There is a server ticket to enable the simplifier for Bonsai https://jira.mongodb.org/browse/SERVER-76509

Reply:
Actually, there is one collection scan test already added, so once Bonsai enabled the simplification we will see the results and let me try to add a big query

Reply:
Added 2 benchmarks as you suggested 

LargeFilterCollectionScan and LargeFilterIndexScan. The simplifier actually managed to simplify them and shows around 2 times better throughput comparing to non-simplified version.

Reply:
Thank you for adding the large filter tests!

Can you also add a smaller collection size version of the LargeFilterCollectionScan workload? (see my first comment for an explanation)

> Actually, there is one collection scan test already added, so once Bonsai enabled the simplification we will see the results and let me try to add a big query

Bonsai (M2) won't run if there are any indexes other than _id (regardless of whether or not the query could use them), which is why I wanted the queries to run before creating indexes (running them after deleting the existing indexes is fine too). Otherwise, there's no point in running this on a Bonsai variant at all.

Could you also add the Bonsai variants to AutoRun after making this change?
```yml
      - standalone-heuristic-bonsai
      - standalone-sampling-bonsai
```

>  The simplifier actually managed to simplify them and shows around 2 times better throughput comparing to non-simplified version.

Very nice! 🎉

Reply:
@alyacb I don't see a way how to run the benchmarks twice on collection of different sizes without copy-pasting everything in some way or another.

Reply:
Actually, I believe that you mean to run only collection scans on smaller workload, which is like 2-3 actors only.

Reply:
Added this second workload [BooleanSimplifierSmallDataset.yml](https://github.com/mongodb/genny/pull/1059#diff-b38d492b33cb9807ed8501874b543718bdd61b8d68a64dadfeb9f1c192e877ea) to test on smaller collections. Also the indexes are created after running collection scans now.